### PR TITLE
OAK-9613 : Define API to retrieve parent or null for a given Item

### PR DESCRIPTION
**Pull request by anchela**

### CODE REVIEW

**Comment by cziegeler** [comment]
Lgtm, thanks!

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java [hunk 1]
```diff
@@ -233,6 +233,35 @@ public Item performNullable() throws RepositoryException {
...
+        }
+    }
+    
+    @NotNull
+    private static ItemImpl checkItemImpl(@NotNull Item item) throws RepositoryException {
+        if (item instanceof ItemImpl) {
+            return (ItemImpl) item;
+        } else {
+            throw new RepositoryException("Invalid item implementation '"+item.getClass().getName()+"'. Excpected org.apache.jackrabbit.oak.jcr.session.ItemImpl");
+        }
```
Comment (reschke) [inline oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java:262]: weird whitespace

**Review by reschke — COMMENTED** [review]
otherwise lgtm

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java (hunk 1 above)
Comment (anchela) [reply oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java:262]: what whitespace is weird? sorry... didn't get it....  but can change it if i know what :)

**Review by smiroslav — APPROVED** [review]

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java (hunk 1 above)
Comment (reschke) [reply oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java:262]: in the arguments for "RepositoryException"; actually *missing* whitespace

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java [hunk 2]
```diff
@@ -233,6 +233,35 @@ public Item performNullable() throws RepositoryException {
...
+            throw new RepositoryException("Item '"+item+"' has been obtained through a different Session.");
+        }
+    }
+    
+    @NotNull
+    private static ItemImpl checkItemImpl(@NotNull Item item) throws RepositoryException {
+        if (item instanceof ItemImpl) {
+            return (ItemImpl) item;
+        } else {
+            throw new RepositoryException("Invalid item implementation '"+item.getClass().getName()+"'. Excpected org.apache.jackrabbit.oak.jcr.session.ItemImpl");
```
Comment (reschke) [inline oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java]:
```suggestion
            throw new RepositoryException("Invalid item implementation '" + item.getClass().getName() + "'. Excpected org.apache.jackrabbit.oak.jcr.session.ItemImpl");
```

Comment (kwin) [reply oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java]: Excpected -> Expected

Comment (mreutegg) [reply oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java]: Use ItemImpl.class.getName() instead of String literal?

Comment (anchela) [reply oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java]: thanks @kwin and @mreutegg , good point.

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java (hunk 1 above)
Comment (anchela) [reply oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java:262]: ah.... yes... will add it :)

**Comment by anchela** [comment]
@reschke . @kwin , @mreutegg , let me know if your findings have been addressed but the subsequent commit.

**Review by mreutegg — APPROVED** [review]
Looks good now.

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java [hunk 3]
```diff
@@ -233,6 +233,35 @@ public Item performNullable() throws RepositoryException {
...
+            throw new RepositoryException("Item '" + item + "' has been obtained through a different Session.");
+        }
+    }
+    
+    @NotNull
+    private static ItemImpl checkItemImpl(@NotNull Item item) throws RepositoryException {
+        if (item instanceof ItemImpl) {
+            return (ItemImpl) item;
+        } else {
+            throw new RepositoryException("Invalid item implementation '" + item.getClass().getName() + "'. Expected " + ItemImpl.class.getName());
```
Comment (reschke) [inline oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java:261]: I'd make that

~~~
"Expected: " + ItemImpl.class.getName() + "."
~~~

But that's just for aesthetics.

**Pull request merged**